Data_Engine/Query/Linq/FilterBy.cs

```diff
@@ -0,0 +1,214 @@
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+
+using BH.oM.Base;
+using BH.oM.Reflection.Attributes;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using BHRE = BH.Engine.Reflection;
+
+namespace BH.Engine.Data
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Filters a list of objects to output only those with a property value matching that input")]
+        [Input("objects", "List of objects to be filtered. All objects in the list should be of similar type")]
+        [Input("propertyName", "The name of the property to filter the objects by")]
+        [Input("value", "The value of the selected property to filter the objects by")]
+        [Input("ignoreStringCase", "Ignore upper/lower case letters if the property filtered by is a text/string.")]
+        [Input("exactStringMatch", "If true checks if the property filtered by contains the text value (filtering must be a string/text)")]
+        public static List<T> FilterBy<T>(this List<T> objects, string propertyName = null, object value = null, bool ignoreStringCase = false, bool exactStringMatch = true)
+        {
+            if (objects == null || objects.Count == 0)
+            {
+                BH.Engine.Reflection.Compute.RecordWarning("No objects submitted to filter");
+                return new List<T>();
+            }
+
+            if (propertyName == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("propertyName cannot be null in order to filter the objects");
+                return new List<T>();
+            }
+
+            if (value == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("value cannot be null to filter the objects");
+                return new List<T>();
+            }
+
+            object firstObject = objects.First(x => x != null);
+            bool isNested = propertyName.Contains(".");
+            bool isCustomData = false;
+
+            System.Reflection.PropertyInfo prop = firstObject.GetType().GetProperty(propertyName);
+            if (prop == null && firstObject is IBHoMObject)
+            {
+                if ((firstObject as IBHoMObject).CustomData.ContainsKey(propertyName))
+                    isCustomData = true;
+            }
+
+            System.Type type = PropertyType(firstObject, propertyName);
+
+            int groupedObjects = objects.GroupBy(x => x.GetType()).Count();
```

Comment:
Ouch! If objects contains null elements, this will throw an exception. 
As I said before, the best is to first do `objects = objects.Where(x => x != null).ToList()`